### DISCUSSION TITLE
OUTLINE USING ONLY 1 CONTAINER (i.e. how to connect to existing Redis & Postgres on localhost, instead of inside new Docker containers)

### DISCUSSION THREAD

**Discussion by ReessKennedy** [outline/outline#3713] [Self Hosting]
## Success
Great work on this software, team! I have successfully installed via Docker, with the help of Portainer. It's working well. [Thomas Griffin's guide was very helpful.](https://thomasgriffin.com/how-to-install-the-outline-knowledge-base-wiki-on-ubuntu/) But ... I have a need to simplify and improve my install and am seeking expert help! 

## Issue: Duplicate Resources
The issue is: On this same server I already have POSTGRES and REDIS running OUTSIDE of Docker and I am using them with other apps running on the server. 

## Desired Solution: Properly connect to existing Postgres and Redis in Docker Compose
To minimize resource usage, I need to figure out how to just pass environment vars in the right way so I can ditch these containers and simplify my server but I haven't been able to properly connect so far, it seems. 

## stack.env (Environment Vars)
In the full Docker version that works three additional vars are added with database info (POSTGRES_USER, POSTGRES_PASSWORD, POSTGRES_DB) but I think these are only needed to pass to the Docker container and if connecting to localhost then just the database URL with all the proper info is sufficient. 

For both the REDIS_URL and DATABASE_URL I have changed location from "postgres:5432" to "localhost:5432" and "redis:6379" to "localhost:6379" just with the thought that this is where it should be connecting and that the use of "postgres:5432" is for connecting to the dockerized postgres location. I have also opened up all these ports in the firewall, made sure they are listening and for the POSTGRES I have created an empty database and username and password and entered them correctly into the DATABASE_URL and made sure the POSTGRES database is accessible by anyone (i.e. not IP restricted) just to make it as liberal as possible in connectivity.

Still, after all this, it doesn't seem to be working ... container appears to be running but actually is occupying port 3000 and this seems to indicate a database connection failure, or some failure. I examine the database I have created and it looks empty of any tables, meaning no build process seems to have taken place. Any ideas GREATLY appreciated!

`
SECRET_KEY=bfc9109f513lqwaksdlkjalkjljklkwlkjlwqe
UTILS_SECRET=1fd7235f8;lkwqel;kqw;;qlk;lkqwwec204133f833fe6c5f9d4
FORCE_HTTPS=false
ENABLE_UPDATES=true
WEB_CONCURRENCY=2

DATABASE_URL=postgres://outlinewikiuser:lkjsalkjlaslkjdlkjasdads@localhost:5432/outlinewiki
PGSSLMODE=disable
REDIS_URL=redis://localhost:6379
URL=https://wiki.mydomain.com
PORT=3000
AWS_REGION=us-east-1
AWS_S3_UPLOAD_BUCKET_NAME=mybucketname
AWS_S3_UPLOAD_BUCKET_URL=https://mybucketname.s3.us-east-1.amazonaws.com
AWS_S3_UPLOAD_MAX_SIZE=26214400
AWS_ACCESS_KEY_ID=Alkjalklksdjfjlkjalsfad
AWS_SECRET_ACCESS_KEY=Zl;kas;lksa;lkd;las;kdk;la;lsk;dlkasdasd
GOOGLE_CLIENT_ID=saljkdlkasjdlkjaslkjdlkjalskjdasd.apps.googleusercontent.com
GOOGLE_CLIENT_SECRET=aksjhkljaslkdjlakjsldjlakjsdas
SLACK_CLIENT_ID=98729782398732985.32p9832009823
SLACK_CLIENT_SECRET=09320983lkjlkjd908089308932
`

## Docker Compose Example 

`version: "3"
services:

  outline:
    image: outlinewiki/outline:latest
    container_name: outlinewiki
    restart: always
    command: sh -c "yarn sequelize:migrate --env=production-ssl-disabled && yarn start --env=production-ssl-disabled"
    ports:
      - "3000:3000"
    env_file:
      - stack.env

**Comment by tommoor** [comment]
```
depends_on:
- postgres
- redis

volumes:
database-data:`
```

All of this should be removed if you're not running them in Docker. What do the logs from `outline` say?

**Comment by ReessKennedy** [comment]
@tommoor! Thank you. Progress has been made: 

## Postgres Works
I got the local Postgres to work by pointing it to the actual IP address of my server instead of using "localhost" because I guess when called from within the docker container, localhost means the localhost of the docker container and not the actual server that container is running on. 

## Redis still having issues 
I altered the link to redis to be something like redis://123.456.780/101:6379 but it won't connect. I also tried setting a password for redis thinking this might be needed if accessing with the IP address but the same thing occured when using  redis://pssowrd@123.456.780/101:6379  and I am running into the following error in the logs ... tough to know how to handle
### Specific Redis Error
{"error":"connect ECONNREFUSED 123.456.789.101:1213","stack":"Error: connect ECONNREFUSED 123.456.789.101:1213\n  at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1157:16)","level":"error","message":"Redis error in socketio adapter"}

Some Googling around this has not immediately helped out.

**Reply by tommoor** [reply]
`1213` is a very unusual Redis port

**Comment by ReessKennedy** [comment]
I figured it out! I had to bind to 0.0.0.0 instead of 127.0.0.1! Thanks for the help here.
